presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -706,6 +706,12 @@ public PlanWithProperties visitValues(ValuesNode node, Context context)
         public PlanWithProperties visitTableCommit(TableCommitNode node, Context context)
         {
             PlanWithProperties child = planChild(node, context.withPreferredProperties(PreferredProperties.any()));
+
+            // if the child is already a gathering exchange, don't add another
+            if ((child.getNode() instanceof ExchangeNode) && ((ExchangeNode) child.getNode()).getType().equals(ExchangeNode.Type.GATHER)) {
```

Comment:
Actually, no. Without this change, we will have a redundant gather inserted into the plan for UNION ALL.